In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
##import seaborn as sns

sales = pd.read_csv("Auto Sales data.csv")

In [4]:
##check the first five records of the dataset
sales.head()

,ORDERNUMBER,QUANTITYORDERED,PRICEEACH,ORDERLINENUMBER,SALES,ORDERDATE,DAYS_SINCE_LASTORDER,STATUS,PRODUCTLINE,MSRP,PRODUCTCODE,CUSTOMERNAME,PHONE,ADDRESSLINE1,CITY,POSTALCODE,COUNTRY,CONTACTLASTNAME,CONTACTFIRSTNAME,DEALSIZE
0,10107,30,95.70,2,2871.00,24/02/2018,828,Shipped,Motorcycles,95,S10_1678,Land of Toys Inc.,2125557818,897 Long Airport Avenue,NYC,10022,USA,Yu,Kwai,Small
1,10121,34,81.35,5,2765.90,07/05/2018,757,Shipped,Motorcycles,95,S10_1678,Reims Collectables,26.47.1555,59 rue de l'Abbaye,Reims,51100,France,Henriot,Paul,Small
2,10134,41,94.74,2,3884.34,01/07/2018,703,Shipped,Motorcycles,95,S10_1678,Lyon Souveniers,+33 1 46 62 7555,27 rue du Colonel Pierre Avia,Paris,75508,France,Da Cunha,Daniel,Medium
3,10145,45,83.26,6,3746.70,25/08/2018,649,Shipped,Motorcycles,95,S10_1678,Toys4GrownUps.com,6265557265,78934 Hillside Dr.,Pasadena,90003,USA,Young,Julie,Medium
4,10168,36,96.66,1,3479.76,28/10/2018,586,Shipped,Motorcycles,95,S10_1678,Technics Stores Inc.,6505556809,9408 Furth Circle,Burlingame,94217,USA,Hirano,Juri,Medium


In [3]:
##check the size of the dataset
print(f"shape: {sales.shape}")


shape: (2747, 20)


In [5]:
##check for the column names
sales.columns

Index(['ORDERNUMBER', 'QUANTITYORDERED', 'PRICEEACH', 'ORDERLINENUMBER',
       'SALES', 'ORDERDATE', 'DAYS_SINCE_LASTORDER', 'STATUS', 'PRODUCTLINE',
       'MSRP', 'PRODUCTCODE', 'CUSTOMERNAME', 'PHONE', 'ADDRESSLINE1', 'CITY',
       'POSTALCODE', 'COUNTRY', 'CONTACTLASTNAME', 'CONTACTFIRSTNAME',
       'DEALSIZE'],
      dtype='str')

In [9]:
##check for the datatypes of the records
sales.dtypes

ORDERNUMBER               int64
QUANTITYORDERED           int64
PRICEEACH               float64
ORDERLINENUMBER           int64
SALES                   float64
ORDERDATE                   str
DAYS_SINCE_LASTORDER      int64
STATUS                      str
PRODUCTLINE                 str
MSRP                      int64
PRODUCTCODE                 str
CUSTOMERNAME                str
PHONE                       str
ADDRESSLINE1                str
CITY                        str
POSTALCODE                  str
COUNTRY                     str
CONTACTLASTNAME             str
CONTACTFIRSTNAME            str
DEALSIZE                    str
dtype: object

In [11]:
##convert the date into a datetime series
sales["ORDERDATE"] = pd.to_datetime(sales["ORDERDATE"], dayfirst = True)


dtype('<M8[us]')

In [12]:
##get the information about the dataset
sales.info()

<class 'pandas.DataFrame'>
RangeIndex: 2747 entries, 0 to 2746
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   ORDERNUMBER           2747 non-null   int64         
 1   QUANTITYORDERED       2747 non-null   int64         
 2   PRICEEACH             2747 non-null   float64       
 3   ORDERLINENUMBER       2747 non-null   int64         
 4   SALES                 2747 non-null   float64       
 5   ORDERDATE             2747 non-null   datetime64[us]
 6   DAYS_SINCE_LASTORDER  2747 non-null   int64         
 7   STATUS                2747 non-null   str           
 8   PRODUCTLINE           2747 non-null   str           
 9   MSRP                  2747 non-null   int64         
 10  PRODUCTCODE           2747 non-null   str           
 11  CUSTOMERNAME          2747 non-null   str           
 12  PHONE                 2747 non-null   str           
 13  ADDRESSLINE1          2747 no

In [13]:
##check for missing values
sales.isnull().sum()

ORDERNUMBER             0
QUANTITYORDERED         0
PRICEEACH               0
ORDERLINENUMBER         0
SALES                   0
ORDERDATE               0
DAYS_SINCE_LASTORDER    0
STATUS                  0
PRODUCTLINE             0
MSRP                    0
PRODUCTCODE             0
CUSTOMERNAME            0
PHONE                   0
ADDRESSLINE1            0
CITY                    0
POSTALCODE              0
COUNTRY                 0
CONTACTLASTNAME         0
CONTACTFIRSTNAME        0
DEALSIZE                0
dtype: int64

In [14]:
##calculate the percentage of missing values in the dataset
percent = (sales["ORDERDATE"].isnull().sum()/len(sales))*100
percent

np.float64(0.0)

In [16]:
##check for duplicate rows
sales.duplicated().sum()

np.int64(0)

In [20]:
##number of un unique items in each columns
categorical_columns = sales.select_dtypes(include ="object").columns
for column in categorical_columns:
    print(f"\n---{column}---")
    print("Unique Values: ",sales[column].nunique())
    print(sales[column].value_counts().head())

<ipython-input-20-fc9d7379f8d5>:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_columns = sales.select_dtypes(include ="object").columns



---STATUS---
Unique Values:  6
STATUS
Shipped      2541
Cancelled      60
Resolved       47
On Hold        44
Name: count, dtype: int64

---PRODUCTLINE---
Unique Values:  7
PRODUCTLINE
Classic Cars    949
Vintage Cars    579
Motorcycles     313
Planes          304
Name: count, dtype: int64

---PRODUCTCODE---
Unique Values:  109
PRODUCTCODE
S18_3232    51
S24_1444    28
S24_2840    28
S32_2509    28
Name: count, dtype: int64

---CUSTOMERNAME---
Unique Values:  89
CUSTOMERNAME
Euro Shopping Channel           259
Mini Gifts Distributors Ltd.    180
Australian Collectors, Co.       55
La Rochelle Gifts                53
Name: count, dtype: int64

---PHONE---
Unique Values:  88
PHONE
(91) 555 94 44    259
4155551450        180
03 9520 4555       55
40.67.8555         53
Name: count, dtype: int64

---ADDRESSLINE1---
Unique Values:  89
ADDRESSLINE1
C/ Moralzarzal, 86              259
5677 Strong St.                 180
636 St Kilda Road                55
67, rue des Cinquante Otages     53
N

In [22]:
##number of records per year
sales["ORDERDATE"].dt.year.value_counts().sort_index

<bound method Series.sort_index of ORDERDATE
2019    1329
2018     953
2020     465
Name: count, dtype: int64>

In [29]:
##validate the data if it is correct
sales["CALCULATED_SALES"] = sales["QUANTITYORDERED"] * sales["PRICEEACH"]
sales["SALES_DIFFERENCE"]=sales["CALCULATED_SALES"]- sales["SALES"]
sales[
    ["QUANTITYORDERED",
     "PRICEEACH",
     "SALES",
     "CALCULATED_SALES",
     "SALES_DIFFERENCE"]
].head(10)

,QUANTITYORDERED,PRICEEACH,SALES,CALCULATED_SALES,SALES_DIFFERENCE
0,30,95.70,2871.00,2871.00,0.000000e+00
1,34,81.35,2765.90,2765.90,-4.547474e-13
2,41,94.74,3884.34,3884.34,-4.547474e-13
3,45,83.26,3746.70,3746.70,4.547474e-13
4,36,96.66,3479.76,3479.76,-4.547474e-13
5,29,86.13,2497.77,2497.77,0.000000e+00
6,48,114.84,5512.32,5512.32,0.000000e+00
7,41,114.84,4708.44,4708.44,9.094947e-13
8,37,107.18,3965.66,3965.66,4.547474e-13
9,23,101.44,2333.12,2333.12,0.000000e+00


In [33]:
##check for anomalies
sales[sales["MSRP"] <= 0]

,ORDERNUMBER,QUANTITYORDERED,PRICEEACH,ORDERLINENUMBER,SALES,ORDERDATE,DAYS_SINCE_LASTORDER,STATUS,PRODUCTLINE,MSRP,...,PHONE,ADDRESSLINE1,CITY,POSTALCODE,COUNTRY,CONTACTLASTNAME,CONTACTFIRSTNAME,DEALSIZE,CALCULATED_SALES,SALES_DIFFERENCE


In [42]:
audit = pd.DataFrame({"column": sales.columns,
                     "Data_type": sales.dtypes.astype(str),
                     "Missing_Values": sales.isnull().sum(),
                     "Missing_percentage": (sales.isnull().sum()/len(sales)*100).round(2),
                     "Unique_values": sales.nunique()
                     }) 
audit

,column,Data_type,Missing_Values,Missing_percentage,Unique_values
ORDERNUMBER,ORDERNUMBER,int64,0,0.0,298
QUANTITYORDERED,QUANTITYORDERED,int64,0,0.0,58
PRICEEACH,PRICEEACH,float64,0,0.0,1843
ORDERLINENUMBER,ORDERLINENUMBER,int64,0,0.0,18
SALES,SALES,float64,0,0.0,2690
ORDERDATE,ORDERDATE,datetime64[us],0,0.0,246
DAYS_SINCE_LASTORDER,DAYS_SINCE_LASTORDER,int64,0,0.0,1832
STATUS,STATUS,str,0,0.0,6
PRODUCTLINE,PRODUCTLINE,str,0,0.0,7
MSRP,MSRP,int64,0,0.0,80


1. Data Audit

Dataset Overview

The dataset contains 2,747 records and 20 variables representing auto sales transactions.

The audit evaluates:
- Dataset structure
- Data types
- Missing values
- Duplicate records
- Categorical variables
- Numerical variables
- Date fields
- Data consistency
- Potential anomalies